[WIP] Refactor planner API

jslee02 opened the pull request:
This is a WIP PR to initiate discussion on introducing planner API.

**Summary**
* Add planner classes
  * `Planner`: Base class
  * `ConcretePlanner : Planner`: Base class for non-meta planners (e.g., `SnapPlanner`). Open to better name.
  * `MetaPlanner`
* Add problem classes (_method_ in `prpy`): This is to encapsulate different sets of parameters per planning problems.

***

**Before creating a pull request**

- [ ] Document new methods and classes
- [x] Format code with `make format`

**Before merging a pull request**

- [x] Set version target by selecting a milestone on the right side
- [ ] Summarize this change in `CHANGELOG.md`
- [x] Add unit test(s) for this change

mkoval commented:
Thanks to @jslee02 for putting together this draft. It's a great start and is very helpful for driving the discussion.

This is actually quite a bit more sophisticated than what I had in mind. I wrote up my original thoughts from back when we started Aikido below. I intend this to be the starting point of further discussion - I am not trying to prescribe this as the only solution. I am very interested in hearing what drove the design decisions in this pull request, since the devil is often in the details of the implementation.

---

I was originally envisioning four "layers" to the planning interface:

1. **The planner itself,** which may comply with no standard interface. The goal of this API is to expose the _full functionality_ of the planner while (a) a providing a clean API, (b) using standard Aikido types wherever possible, and (c) introducing minimal overhead. Often, this will be as simple as a free function. All of the planning functions in Aikido today fall into this layer, e.g. `planOMPL()`, `planCRRTConnect()`, and `computeParabolicTiming()`.

2. **A planner interface** for each type of query. Instead of having a singular `Planner` interface, there would be a separate interface for each type of query. We would define a new interface if: (a) the query is common in our application code (e.g. MAGI actions) and (b) more than one planner can solve the query, i.e. it is possible to implement more than one adapter class as outlined in (3) below.

3. **An adapter class,** that uses (1) to solve queries specified in a standard format. Any arguments that are not part of the planner interface, e.g. planner-specific resolution parameters, should be specified by the adapter's constructor. Often, this involves converting the query into the type of constraints that the planner accepts. For example, an adapter class could solve a "plan to a goal configuration" query by calling `planOMPL()` with a `FiniteSampleable` goal constraint.

4. **A meta-planner** implements an adapter class using another planner interface, i.e. it converts from one planner interface to another. For example, a meta-planner may convert an arbitrary `PlanToConfigurations` planner to a `PlanToConfiguration` planner by calling `PlanToConfigurations` on a list that contains the singular goal configuration.

If we need a meta planner that operates on multiple types of planning queries, e.g. by calling multiple planners in sequence, then we can implement that as a template class. The details here are are a little hairy, so we need to think carefully about how we define the planner interfaces to make this possible.

---

These are the general principles that drove this design:

- It should always be possible to access the full power of a planner. Implementing an interface should never reduce a planner's functionality.
- An interface only has value if all implementations of the interface can be used interchangeably (sometimes called the [Liskov substitution principle](https://en.wikipedia.org/wiki/Liskov_substitution_principle)).
- Every planning query has a different interface, so it's not possible to use them interchangeably.
- It's useful to enforce the type on arguments and return values of planning functions at compile time.
- Trying to implement a generic `Planner` base class requires bypassing most of the C++ type system. If that is necessary, you should consider writing that code in Python instead. :winking:

Looking at the current implementation in this pull request, I think there are a few great ideas that we could add to the mix:

- **`Problem` classes** provide a nice mechanism for defining arguments. Inheritance could come in handy here once we have a wider variety of planner interfaces. For example, we could write helper code for a lot of boilerplate `StateSpace`, `DistanceMetric`, etc initialization. It also is a nice workaround for C++'s lack of named initialization lists.
- **`Result` classes** provide a nice mechanism for annotating the output of planners with planner-specific information. This has many of the same advantages as `Outcome` did in https://github.com/personalrobotics/aikido/pull/266.

---

Here is a more concrete example in C++-like pseudocode:

```c++
// 1. Planner
trajectory::InterpolatedUniquePtr planOMPL(
  const statespace::StateSpace::State* _start,
  const statespace::StateSpace::State* _goal,
  statespace::StateSpacePtr _stateSpace,
  statespace::InterpolatorPtr _interpolator,
  distance::DistanceMetricPtr _dmetric,
  constraint::SampleablePtr _sampler,
  constraint::TestablePtr _validityConstraint,
  constraint::TestablePtr _boundsConstraint,
  constraint::ProjectablePtr _boundsProjector,
  double _maxPlanTime,
  double _maxDistanceBtwValidityChecks);

// 2. Planner Interfaces
class FromConfigurationToConfigurationPlanner
{
public:
  virtual trajectory::TrajectoryUniquePtr planFromConfigurationToConfiguration(
    const statespace::StateSpacePtr& stateSpace,
    const statespace::StateSpace::State* startState,
    const statespace::StateSpace::State* goalState) = 0;

};

class FromConfigurationToConfigurationsPlanner
{
public:
  virtual trajectory::TrajectoryUniquePtr planFromConfigurationToConfigurations(
    const statespace::StateSpacePtr& stateSpace,
    const statespace::StateSpace::State* startState,
    std::vector<const statespace::StateSpace::State*> goalStates) = 0;
};


// 3. Adapter Interfaces
class OMPLFromConfigurationsToConfigurationsPlanner
  : public virtual FromConfigurationToConfigurationsPlanner
{
public:
  virtual trajectory::TrajectoryUniquePtr planFromConfigurationToConfigurations(
    const statespace::StateSpacePtr& stateSpace,
    const statespace::StateSpace::State* startState,
    std::vector<const statespace::StateSpace::State*> goalStates)
  {
    return planOMPL(...);
  }
};


// 4. Meta-Planner
class FromConfigurationsToConfigurations_to_FromConfigurationToConfiguration
  : public virtual FromConfigurationToConfigurationPlanner
{
public:
  explicit FromConfigurationsToConfigurations_to_FromConfigurationToConfiguration(
      std::unique_ptr<OMPLFromConfigurationsToConfigurationsPlanner> delegatePlanner)
    : mDelegatePlanner{std::move(delegatePlanner)}
  {
  }

  virtual trajectory::TrajectoryUniquePtr planFromConfigurationToConfiguration(
    const statespace::StateSpacePtr& stateSpace,
    const statespace::StateSpace::State* startState,
    const statespace::StateSpace::State* goalState)
  {
    return mDelegatePlanner->planFromConfigurationToConfigurations(
      stateSpace, startState, {goalState});
  }
};

```

jslee02 commented:
@mkoval Thanks for the suggestion. I think the major differences between our designs are:
(1) Having base planner class (i.e., `Planner`) vs. not
(2) Having all the solvable planning methods in a planner vs. having them as separate classes

More specifically, I propose having a planner class and putting all the functions of the solvable planning problems (methods) into the planner class. So we might only have like `SnapPlanner`, `VectorFieldPlanner`, `OmplPlanner`, `TrajOptPlanner`, and so on; and multiple planning functions will be in each class. In your design, we would have more classes which are the permutation of planners and planning problems, and single planning function in each class.

Here are some thoughts on possible advantages of this PR's implementation that I believe:

Having planning functions in a specific planner class would be able to share reusable functionality and data in the same class. I believe those planning functions in a planner class have a similar fashion of solving the problems. For example, `TrajOptPlanner` would solve all the problems solving optimization problems. It would be still possible with your design by having an additional class(es) and composing it into the necessary adaptor classes, but I think this could introduce unnecessary complex (I could be biased).

Another possible advantage is that the base planner class can provide useful interface such as `canSolve(problem)`, `solve(problem)`, and `getSolvableProblems()`. These functions allow you to get the capability of the planner. This functionality would be able to be achieved with your design by adding them to meta-planner, but my proposal would enable to implement meta-planners straightforwardly with the virtual functions of the base planner class.

I'm not sure how to incorporate `Problem` classes into your design because it seems the planner interfaces are the problem classes to me. Probably we have different definitions of `Problem` classes in each mind? :sweat_smile:

mkoval commented:
I think your summary of the design differences is accurate. 👍 

---

> In your design, we would have more classes which are the permutation of planners and planning problems, **and [a] single planning function in each class.**

This is not strictly required. If the implementation of multiple planning interfaces for a planner are similar, then you can implement them in the same class. However, I do think that this will end up being relatively rare in practice - I think an adapter class is often the better way to go.

> So we might only have like `SnapPlanner`, `VectorFieldPlanner`, `OmplPlanner`, `TrajOptPlanner`, and so on; and multiple planning functions will be in each class.

This is the approach that we used in PrPy. We quickly discovered that it is difficult to handle planner-specific parameters. Multiple planners may implement the same planning method (e.g. `PlanToConfiguration`), but take different parameters. For example: some planners require time limits, others require heuristics, etc.

With my proposed API, planner-specific parameters (e.g. time limits) are passed into the adapter class' constructor. The planning method itself always take only the arguments defined by the planner interface(s) that it implements.

This is critical when planners are used via dependency injection. For example, a MAGI action can take an `std::unique_ptr<PlanToEndEffectorOffset>` without caring which specific planner it is using or what extra `**kwargs` it needs to pass to it.

> Having planning functions in a specific planner class would be able to share reusable functionality and data in the same class. [...] **It would be still possible with your design by having an additional class(es)** and composing it into the necessary adaptor classes, but I think this could introduce unnecessary complex (I could be biased).

This is exactly how I propose to solve the re-use problem. Nothing about defining a variety of interfaces prohibits us from defining helper functions and/or classes. I suspect that the resulting code would end up being more reusable and easier to unit test, since it separates the helper code from the implementation of the planner. (This is largely my gut reaction, so I don't want to start a holy war. 😬)

> Another possible advantage is that the base planner class can provide useful interface such as `canSolve(problem)`, `solve(problem)`, and `getSolvableProblems()`. These functions allow you to get the capability of the planner. 

This is a neat idea. This seems most useful for allowing a `MetaPlanner` to have the upmost flexibility in deciding which planner to apply to a problem. The disadvantage is that you lose a lot of nice type safety enforced by the C++ type system. My approach is much more rigid because the list of queries is specified as C++ types at compile time, but has that type safety.

While it has a lot of potential, I am not sure how useful this level of generality will be in practice today. We have relatively small sets of planners and types of planning queries. However, I do entirely understand the appeal of this - I believe this is actually pretty similar to the design that @cdellin and the NREC folks used for motion planning on CHIMP in the DRC.

---

In general, I was trying to follow the philosophy that Aikido is a toolbox of small, reusable pieces of code that can stitched together to solve a variety of planning problems. I really like the idea of adapter classes because it is so [DRY](https://en.wikipedia.org/wiki/Don%27t_repeat_yourself). Uou can write (and test!) a piece of code once, rather than repeating it once per planner.

For example, any planner that implements `PlanFromConfigurationsToConfigurations` would automatically get `PlanToConfigurationFromConfigurations`, `PlanFromConfigurationsToConfiguration`, and `PlanToConfigurationToConfiguration` for free. You could also convert any planner that implements `PlanToConfiguration` to `PlanToTSR` by sampling from the TSR and calling `PlanToConfiguration` (this is what `TSRPlanner` does in PrPy).

This is similar to how we defined `NewtonsMethodProjectable` once, rather than re-implementing a rootfinder every time we need to do constraint projection.

---

I am interested in hearing your thoughts on this @jslee02 - I'm open to be convinced otherwise. 😈

siddhss5 commented:
At a meta-level, we seem to be discussing the relative merits of composition over inheritance. @mkoval and @jslee02 are way better at discussing this than I am, so I'll defer to them, except point out that composition often enables more flexibility whereas inheritance often enables more readability.

I want to emphasize one comment that @mkoval made that I think is critical for any path we pursue:

> This is the approach that we used in PrPy. We quickly discovered that it is difficult to handle planner-specific parameters. Multiple planners may implement the same planning method (e.g. PlanToConfiguration), but take different parameters. For example: some planners require time limits, others require heuristics, etc.

This has annoyed me [maybe us all] no end. Although I am a fan of having some structure in what a planner can do [and I think the `PlanTo*` structure we came up with in `prpy` is really fantastic], I always felt like the way it was implemented in `prpy` really really straightjacketed us in accessing the planners. I suppose one way to address this is how `scipy` and much of `MATLAB` address this is via an open-ended `options` dictionary. But then that demands a _lot_ of documentation and can easily go very wrong.

gilwoolee commented:
Hi all, thanks for the great suggestions. I think both designs are flexible and capable to handle the variety of problems we have. In terms of readability, I slightly prefer @jslee02 's proposal. However, I think there's one major advantage of @mkoval 's proposal that may be a bit complicated to handle in @jslee02 's current design:

> With my proposed API, planner-specific parameters (e.g. time limits) are passed into the adapter class' constructor. The planning method itself always take only the arguments defined by the planner interface(s) that it implements. This is critical when planners are used via dependency injection. For example, a MAGI action can take an std::unique_ptr<PlanToEndEffectorOffset> without caring which specific planner it is using or what extra **kwargs it needs to pass to it.

These are in @jslee02 's version `problem x planner`-specific parameters, so it cannot be passed in at the constructor of @jslee02 's planner (e.g. `OMPLPlanner`). @jslee02 and I had a brief discussion on how to handle this, e.g. via DART's Aspect functionality, which (to my best understanding) is a sophisticated version of an open-ended `options`-dictionary.  The amount of documentation may not be too different from what we may end up doing with @mkoval 's design, but given that we have to take such an approach, I think being able to rely on C++'s type safety is more appealing.

jslee02 commented:
> This is the approach that we used in PrPy. We quickly discovered that it is difficult to handle planner-specific parameters. Multiple planners may implement the same planning method (e.g. PlanToConfiguration), but take different parameters. For example: some planners require time limits, others require heuristics, etc.

As @gilwoolee mentioned, we could use DART's aspect mechanism (of @mxgrey) to achieve that. In our case, `[~]Problem` will be `Composite`s, and each additional planner specific parameters will be in `Aspect`s, which will be defined in each planner. As so, the planner can use the aspect when it's added to the problem; otherwise, use the defaults. I believe `Aspect`-`Component` is designed to handle this kind of problems.

> (@mkoval's comment) In general, I was trying to follow the philosophy that Aikido is a toolbox of small, reusable pieces of code that can stitched together to solve a variety of planning problems. I really like the idea of adapter classes because it is so DRY. Uou can write (and test!) a piece of code once, rather than repeating it once per planner.
> (@siddhss5's comment) At a meta-level, we seem to be discussing the relative merits of composition over inheritance. @mkoval and @jslee02 are way better at discussing this than I am, so I'll defer to them, except point out that composition often enables more flexibility whereas inheritance often enables more readability.

I also prefer DRY, and it can be achieved in my design as well. `[X]Problem` class could have reusable common functionalities, which are the X planning problem specific. I believe the problem classes are the same with the planner interface in your pseudo code in that sense; except (1) the planning functions are defined in planner classes in my design, and (2) problem classes are derived as adapter planners for each planning methods in your design.

Overall, I believe both designs don't have any major technical issue to achieve certain flexibility we want to obtain, but I think my design would be more straightforward to implement the planner capability checking functionality (e.g., `canSolve(problem)`, `solve(problem)`, and `getSolvableProblems()`).

I think it would be helpful to compare and improve the design ideas here (or in gist). Let me write down sample code soon. :)

jslee02 commented:
@brianhou, @gilwoolee, and I discussed further, and I think we can reach a compromise. Essentially, it would be based on your _Planner Interfaces_ and incorporate `Problem` classes and `canSolve`/`getSolveableProblems` into that.

Here is a high-level picture I have:

**Summary**
* Define planner interfaces per planning problems
* Define adapter interfaces
* Pass `Problem` classes to the planner interfaces
* Add `canSolve` and `getSolveableProblems` to `Robot` class
* Make a concrete robot class (e.g., `Herb`) inherit `Robot` class and planner interfaces that the robot want to support

Please let me know what you guys think or what is not clear. If this sounds good, I will make the changes in this PR.

jslee02 commented:
I've revised the planner API according to the discussion (+ ideas came up with while implementing it). Let me know what do you think. Here is a rough summary of the changes.

**Note on the Changes**
- Kept `Planner` class to provide `canPlan`, `getPlannableProblems`, and `plan`, which takes the base problem class
- Added `PlannerForSingleProblem : Planner` which is a base class of _the planner interface_ classes. It is templatized for a planning problem and implements all the virtual function of `Planner` class.
- Added `PlannerForMultiProblem : Planner` which is the base class of concrete planner class that can solve more than one planning problem
- Revised snap planner (i.e., `SnapConfigurationToConfigurationPlanner`) to inherit `PlannerForSingleProblem`
- Renamed planning problem classes in this fashion: <start_type>To<goal_type>. For example, `ConfigurationToConfiguration`
- Renamed planner classes in the same fashion of planning problem but with '-Planner` appendix.

jslee02 commented:
@mkoval @psigen Could you guys take a look the latest changes if it makes sense when you have a chance? :smile: Thanks!

gilwoolee commented on include/aikido/planner/SnapConfigurationToConfigurationPlanner.hpp:
```diff
@@ -0,0 +1,44 @@
+namespace aikido {
+namespace planner {
+
+/// Planner that checks the straight-line trajectory to the goal.
```
`checks` -> `plans`

brianhou commented on include/aikido/planner/ConfigurationToEndEffectorOffset.hpp:
```diff
@@ -0,0 +1,85 @@
+  /// Constructor.
+  ///
+  /// \param stateSpace State space.
+  /// \param bodyNode Body Node or robot for which the path is to be planned.
```
Is this code written this way because we think it'd be valuable for a robot (e.g. RACECAR) that doesn't have an end-effector to be able to plan to an offset in its current direction?

aditya-vk replied:
Maybe change the name of the problem also in that case? End Effector is very much manipulator specific. We'd like the planners and problems to be generic no? Maybe `SingleSourceSingleTargetFixedOrientationProblem`. Took me a while to type that name. :P [Still not a good enough name though]

jslee02 replied:
I think this is just about confusing description. I renamed it to `endEffectorBodyNode` (and getters and setters accordingly) and improved the descriptions.

aditya-vk commented on include/aikido/planner/ConfigurationToConfiguration.hpp:
```diff
@@ -0,0 +1,85 @@
+  /// Returns the name of the planner problem.
+  static const std::string& getStaticName();
+
+  /// Sets the start state.
```
For Single Source Single Target, I agree that we _could_ create new problems. I don't see any problem with reusing by changing the start and goal though. For instance, at least the current implementation of `planToConfigurations` under `robot/util.cpp` is just iterating through each goal state and calling `planToConfiguration(goal)`. If we are doing that here as well, we could just `setGoalState` instead of creating a new instance everytime.

aditya-vk commented on include/aikido/planner/ConfigurationToConfiguration.hpp:
```diff
@@ -0,0 +1,85 @@
+  const statespace::StateSpace::State* getStartState() const;
+
+  /// Sets the goal state.
+  void setGoalState(const statespace::StateSpace::State* state);
```
Same as above.

aditya-vk commented on include/aikido/planner/ConfigurationToConfigurationPlanner.hpp:
```diff
@@ -0,0 +1,28 @@
+  using TheProblem = ConfigurationToConfiguration;
+  using TheResult = Result;
+
+  virtual trajectory::TrajectoryPtr plan(
```
Should we have `SingleSourceSingleTargetPlanner`, `SingleSourceMultipleTargetsPlanner` etc.?
Similarly for problems. The fact that the planning problem/planner names differ by a single letter is super confusing. No?

jslee02 replied:
It was to differentiate the concrete problem the planner can solver from `Problem` (base class). Renamed to `SolverbleProblem`. Open to here better name. :smile:

brianhou replied:
What about `ProblemType`?

jslee02 replied:
@brianhou I prefer `SolvableProblem` because it's more descriptive, but I'm also open to changing it to `ProblemType`.

aditya-vk commented on include/aikido/planner/ConfigurationToConfigurations.hpp:
```diff
@@ -0,0 +1,73 @@
+  /// \param goalStates Goal states.
+  /// \param interpolator Interpolator used to produce the output trajectory.
+  /// \param constraint Trajectory-wide constraint that must be satisfied.
+  /// \throw If \c stateSpace is not compatible to \c constraint's state space.
```
nit: compatible with?

aditya-vk commented on include/aikido/planner/ConfigurationToConfigurations.hpp:
```diff
@@ -0,0 +1,73 @@
+  /// Returns the name of the planner problem.
+  const std::string& getName() const override;
+  static const std::string& getStaticName();
+
```
What about `addGoal()` method?

aditya-vk commented on include/aikido/planner/ConfigurationToConfigurations.hpp:
```diff
@@ -0,0 +1,73 @@
+  ConfigurationToConfigurations(
+      statespace::StateSpacePtr stateSpace,
+      const statespace::StateSpace::State* startState,
+      const std::vector<statespace::StateSpace::State*>& goalStates,
```
I agree with @brianhou that we would need a set for fast lookup. From a performance perspective, the whole point of having this feature, to my understanding, is to parallelly search for multiple goals and terminate when path to one of them is found. This implies, we would ideally not want to _iterate_ through each goal and plan. I think if done right, this would need a set and not a vector. Correct me if I am wrong.

jslee02 replied:
Agreed. Changed to `std::unordered_set` instead of `std::set` because the order doesn't matter here.

aditya-vk commented on include/aikido/planner/ConfigurationToConfigurations.hpp:
```diff
@@ -0,0 +1,73 @@
+  constraint::TestablePtr getConstraint();
+
+  /// Returns the constraint that must be satisfied throughout the trajectory.
+  constraint::ConstTestablePtr getConstraint() const;
```
Can we have this in `PlanConfigurationToConfiguration` problem as well.

aditya-vk commented on include/aikido/planner/ConfigurationToConfigurations.hpp:
```diff
@@ -0,0 +1,73 @@
+  statespace::InterpolatorPtr mInterpolator;
+
+  /// Trajectory-wide constrained that must be satisfied.
+  constraint::TestablePtr mConstraint;
```
constraint should be `const` during planning.

aditya-vk commented on include/aikido/planner/ConfigurationToConfigurations.hpp:
```diff
@@ -0,0 +1,73 @@
+  const statespace::StateSpace::State* mStartState;
+
+  /// Goal States.
+  const std::vector<statespace::StateSpace::State*> mGoalStates;
```
If we have an `addGoal()` this should not be `const`. Instead it should be `std::vector<const --> mGoalStates` which should be the case nevertheless.

aditya-vk commented on include/aikido/planner/ConfigurationToEndEffectorOffset.hpp:
```diff
@@ -0,0 +1,85 @@
+
+  /// Returns the start state.
+  const statespace::StateSpace::State* getStartState() const;
+
```
Like @brianhou mentioned, none of the problems except for `ConfigurationToConfiguration` have setters for start and goal.

aditya-vk commented on include/aikido/planner/ConfigurationToEndEffectorOffset.hpp:
```diff
@@ -0,0 +1,85 @@
+  statespace::InterpolatorPtr mInterpolator;
+
+  /// Trajectory-wide constraint that must be satisfied.
+  constraint::TestablePtr mConstraint;
```
Constraint should be `const`

aditya-vk commented on include/aikido/planner/ConfigurationToEndEffectorPose.hpp:
```diff
@@ -0,0 +1,84 @@
+  statespace::InterpolatorPtr mInterpolator;
+
+  /// Trajectory-wide constraint that must be satisfied.
+  constraint::TestablePtr mConstraint;
```
As before.

aditya-vk commented on include/aikido/planner/ConfigurationToTSR.hpp:
```diff
@@ -0,0 +1,72 @@
+  statespace::InterpolatorPtr mInterpolator;
+
+  /// Trajectory-wide constraint that must be satisfied.
+  constraint::TestablePtr mConstraint;
```
As before.

aditya-vk commented on include/aikido/planner/Planner.hpp:
```diff
@@ -0,0 +1,55 @@
+  class Result;
+
+  /// Returns true if this planner can solve \c problem.
+  virtual bool canPlan(const Problem* problem) const = 0;
```
I agree.

aditya-vk commented on src/planner/ConfigurationToConfiguration.cpp:
```diff
@@ -0,0 +1,95 @@
+}
+
+//==============================================================================
+constraint::TestablePtr ConfigurationToConfiguration::getConstraint() const
```
Repeated code. Can we have this in the base problem class instead? All problems seem to be sharing the same implementations. At least the const versions can be implemented in the base class.

jslee02 replied:
It sounds good. I'll move this function to the base then.

aditya-vk commented on src/planner/ConfigurationToConfigurations.cpp:
```diff
@@ -0,0 +1,82 @@
+}
+
+//==============================================================================
+const std::vector<statespace::StateSpace::State*>
```
`const std::vector<const statespace::StateSpace::State*>`

Although like I mentioned, we would want an `addGoal()` so the container cannot be `const` in that case.

aditya-vk commented on src/planner/ConfigurationToConfigurations.cpp:
```diff
@@ -0,0 +1,82 @@
+{
+  return mConstraint;
+}
+
```
DRY.

jslee02 commented on include/aikido/planner/CompositePlanner.hpp:
```diff
@@ -0,0 +1,62 @@
+  ///
+  /// \throw If index is equal to or greater than the number of planners in this
+  /// CompositePlanner.
+  PlannerPtr getPlanner(std::size_t index);
```
Your inference is correct. Let me add `getNumPlanners()`.

One additional rationale behind this is for const-correctness (again :laughing:). Since this class already has `getPlanners()`, it can be seemed `getPlanner()` is not necessary. But, `getPlanners()` should be non-const because it returns `std::vector<PlannerPtr>`, which is a list of non-const instances. We can't call this function from const instance of `CompositePlanner` unless we additionally maintain `std::vector<ConstPlannerPtr>`. This is a considerable approach, but for now, let's have const/non-const `getPlanner()` for simplicity.

brianhou replied:
Sorry, I think I'm still missing something. Are you saying that `getPlanner` and `getNumPlanners` are necessary in order for a meta-planner to work? (And we aren't able to just iterate over the vector of planners because of this const-correctness issue? Why can't we just always have (one) vector of `ConstPlanner`s? Maybe a related question: when would we want to change a `Planner`?)

aditya-vk replied:
There _can_ be instances when the planner is changing its parameters as it plans. If these params are modified as member variables, we might need `<non-const>Planner`. :thinking:

jslee02 replied:
@aditya-vk I also considered that case at first, but now I think the instances being changed as it plans should be mutable. This is because I think planner logically shouldn't need to change its internal states but just the returning trajectory and the result.

So I'm thinking of having const planners, and then having only `getPlanners()` should be okay.

jslee02 replied:
Since we decided not to make `plan()` const, this function shouldn't be removed for these reasons:
* `getPlanner(index) const` for const iteration
* `getPlanner(index)` to provide consistent API with `getPlanner(index) const`.

But I'm open to removing `getPlanner(index)` because we could still do non-const iteration using `getPlanners()`.

brianhou replied:
If I remember our discussion correctly, I think we concluded that external users might not need to be able to iterate over the individual planners at all? I'm in favor of removing `getPlanner(index)` and `getNumPlanners`, and potentially adding them back if we find they're useful or necessary.

jslee02 commented on include/aikido/planner/CompositePlanner.hpp:
```diff
@@ -0,0 +1,62 @@
+  virtual bool canPlan(const Problem* problem) const override;
+
+  // Documentation inherited.
+  std::unordered_set<std::string> getPlannableProblems() const override;
```
Removed this function entirely.

jslee02 commented on src/planner/CompositePlanner.cpp:
```diff
@@ -0,0 +1,118 @@
+  if (std::find(mPlanners.begin(), mPlanners.end(), planner) != mPlanners.end())
+  {
+    dtwarn << "[CompositePlanner::addPlanner] Attempts to add a planner that "
+              "already exists in this CompositePlanner.\n";
```
[`std::endl` is `\n` plus flushing](), which an unnecessary overhead I think.

jslee02 commented on src/planner/PlanningResult.cpp:
```diff
@@ -0,0 +1,21 @@
+#include "aikido/planner/PlanningResult.hpp"
```
That's true. This class will be removed once all the old planning functions that use this class are removed or converted to the new API.

I'll add a note to this class about this.

jslee02 commented on src/planner/ConfigurationToConfiguration.cpp:
```diff
@@ -0,0 +1,95 @@
+//==============================================================================
+const std::string& ConfigurationToConfiguration::getStaticName()
+{
+  static std::string name("ConfigurationToConfiguration");
```
Since this is a static function, it cannot return a local variable.

jslee02 commented on include/aikido/planner/PlanningResult.hpp:
```diff
@@ -6,9 +6,20 @@
 namespace planner {
 
-struct PlanningResult
+class PlanningResult
```
Yes. Let's remove once all the old planning API uses this class is removed.

jslee02 commented on include/aikido/planner/ConfigurationToConfiguration.hpp:
```diff
@@ -0,0 +1,85 @@
+
+  /// Returns the constraint that must be satisfied throughout the trajectory.
+  constraint::TestablePtr getConstraint() const;
+  // TODO: Should return ConstInterpolatorPtr when resolving const correctness
```
Removed this TODO as the issue is resolved.

jslee02 commented:
I have revised this PR based on the online and offline discussions. @brianhou, @gilwoolee, @aditya-vk Second round of review, please. :smile:

jslee02 commented on include/aikido/planner/ConfigurationToEndEffectorOffset.hpp:
```diff
@@ -0,0 +1,75 @@
+      dart::dynamics::ConstBodyNodePtr endEffectorBodyNode,
+      const statespace::StateSpace::State* startState,
+      const Eigen::Vector3d& direction,
+      double distance,
```
Passing const value is not preferred for [this reason](https://stackoverflow.com/a/46292715):

jslee02 commented on include/aikido/planner/RankedMetaPlanner.hpp:
```diff
@@ -0,0 +1,20 @@
+public:
+  // Documentation inherited.
+  trajectory::TrajectoryPtr plan(
+      const Problem& problem, Result* result = nullptr) override;
```
We can mark `final` specifier here when we're sure any derived class don't override the virtual function. I just leave it open because we haven't implemented this class yet.

jslee02 commented on include/aikido/statespace/GeodesicInterpolator.hpp:
```diff
@@ -15,12 +15,12 @@ class GeodesicInterpolator : public Interpolator
   ///
   /// \param _stateSpace used for interpolation
-  explicit GeodesicInterpolator(statespace::StateSpacePtr _stateSpace);
+  explicit GeodesicInterpolator(statespace::ConstStateSpacePtr _stateSpace);
```
That's true. I just wouldn't prefer including those fixes in this PR to make this PR as small as possible.

gilwoolee commented on include/aikido/planner/Problem.hpp:
```diff
@@ -0,0 +1,50 @@
+  statespace::ConstStateSpacePtr getStateSpace() const;
+
+  /// Returns the constraint that must be satisfied throughout the trajectory.
+  constraint::ConstTestablePtr getConstraint() const;
```
If the problem was created with nullptr, does it return a `SatisfiedPtr` or a nullptr?

jslee02 replied:
It returns `SatisfiedPtr`. Added note about this to the constructor.

gilwoolee commented on include/aikido/planner/SequenceMetaPlanner.hpp:
```diff
@@ -0,0 +1,20 @@
+namespace aikido {
+namespace planner {
+
+class SequenceMetaPlanner : public CompositePlanner
```
Class docstring? Is this for using a list of planners to solve one problem?

jslee02 replied:
This class is to solve multiple problems. If we want to have a composite planner to solve only one problem, then we could have
* `SingleProblemCompositePlanner<ProblemType>`
* `MultiProblemCompositePlanner` that is renamed from `CompositePlanner`.

gilwoolee commented on include/aikido/planner/ConfigurationToConfigurationPlanner.hpp:
```diff
@@ -0,0 +1,32 @@
+      statespace::ConstStateSpacePtr stateSpace);
+
+  /// Solves \c problem returning the result to \c result.
+  virtual trajectory::TrajectoryPtr plan(
```
Can we have docstring for the params?

gilwoolee commented on include/aikido/planner/ConfigurationToEndEffectorOffset.hpp:
```diff
@@ -0,0 +1,75 @@
+  /// offest while maintaining the current orientation.
+  /// \param[in] startState Start state.
+  /// \param[in] direction Direction of motion [unit vector in the world frame].
+  /// \param[in] distance Distance to move, in meters.
```
Let's also mention that this must be nonnegative.

brianhou commented on include/aikido/planner/ConfigurationToConfigurations.hpp:
```diff
@@ -0,0 +1,62 @@
+#ifndef AIKIDO_PLANNER_PLANTOCONFIGURATIONS_HPP_
```
Nit: `CONFIGURATIONTOCONFIGURATIONS`

brianhou commented on include/aikido/planner/ConfigurationToEndEffectorOffset.hpp:
```diff
@@ -0,0 +1,77 @@
+#ifndef AIKIDO_PLANNER_PLANTOENDEFFECTOROFFSET_HPP_
```
Nit: `CONFIGURATIONTOENDEFFECTOROFFSET`.

brianhou commented on include/aikido/planner/ConfigurationToEndEffectorPose.hpp:
```diff
@@ -0,0 +1,64 @@
+#ifndef AIKIDO_PLANNER_PLANTOENDEFFECTORPOSE_HPP_
```
Nit: `CONFIGURATIONTOENDEFFECTORPOSE`.

brianhou commented on include/aikido/planner/ConfigurationToTSR.hpp:
```diff
@@ -0,0 +1,63 @@
+#ifndef AIKIDO_PLANNER_PLANTOTSR_HPP_
```
Nit: `CONFIGURATIONTOTSR`.

brianhou commented on src/planner/ConfigurationToEndEffectorOffset.cpp:
```diff
@@ -0,0 +1,69 @@
+  if (direction.norm() < 1e-6)
+    throw std::invalid_argument("Length of direction shouldn't be zero.");
+
+  if (distance < 0.0)
```
I personally prefer to be allowed to pass in a negative distance. We can flip the direction and distance in here, if our underlying planners would be unhappy with that.

brianhou commented on src/planner/ConfigurationToEndEffectorOffset.cpp:
```diff
@@ -0,0 +1,69 @@
+  , mDirection(direction.normalized())
+  , mDistance(distance)
+{
+  if (direction.norm() < 1e-6)
```
What happens in line 19 if the norm is zero? If it can be successfully normalized, I think we should just allow it.

jslee02 replied:
According to the [Eigen documentation](https://eigen.tuxfamily.org/dox/classEigen_1_1MatrixBase.html#ad16303c47ba36f7a41ea264cb26bceb6), zero vector will be unchanged. Otherwise, it'll be successfully normalized. So I just changed this to compare to zero instead of `1e-6`.

If we allow zero vector, then I think we should document this as well.

jslee02 commented:
The build failures on macOS should be fixed once #395 is merged.

gilwoolee commented on include/aikido/planner/ConfigurationToEndEffectorOffset.hpp:
```diff
@@ -0,0 +1,79 @@
+#ifndef AIKIDO_PLANNER_CONFIGURATIONTOCONFIGURATIONS_HPP_
```
SHould be `AIKIDO_PLANNER_CONFIGURATIONTOENDEFFECTOROFFSET_HPP_`

gilwoolee commented on include/aikido/planner/ConfigurationToEndEffectorOffset.hpp:
```diff
@@ -0,0 +1,79 @@
+  /// \param[in] interpolator Interpolator used to produce the output
+  /// trajectory.
+  /// \param[in] constraint Trajectory-wide constraint that must be satisfied.
+  /// \throw (1) If \c stateSpace is not compatible with \c constraint's state
```
Why do we need (1)?

gilwoolee commented on include/aikido/planner/ConfigurationToEndEffectorOffset.hpp:
```diff
@@ -0,0 +1,79 @@
+      dart::dynamics::ConstBodyNodePtr endEffectorBodyNode,
+      const statespace::StateSpace::State* startState,
+      const Eigen::Vector3d& direction,
+      double signedDistance,
```
Does this mean that it can have both positive and negative? (We'll have to make sure this is supported in underlying planners.)

jslee02 replied:
Yes. If the underlying planners don't support negative distance, they could flip the direction and sign or throw an exception if flipping doesn't make sense to the planner.

gilwoolee commented on src/planner/CompositePlanner.cpp:
```diff
@@ -0,0 +1,43 @@
+}
+
+//==============================================================================
+bool CompositePlanner::hasPlanner(const Planner* planner) const
```
Could it be more useful to do a type check instead of reference check?

jslee02 replied:
This class is general to handle the case of having different planner instances of the same type. In that case, type checking will return the first planner instance of the type. Or we could have another function that returns set of planners of the type.

As a side note, it'd be better to take type for type checking.

brianhou commented on tests/planner/test_SnapPlanner.cpp:
```diff
@@ -51,43 +57,80 @@ class SnapPlannerTest : public ::testing::Test
   shared_ptr<FailingConstraint> failingConstraint;
   shared_ptr<GeodesicInterpolator> interpolator;
-  aikido::planner::PlanningResult planningResult;
+  SnapConfigurationToConfigurationPlanner::Result planningResult;
```
What's the difference between `SnapConfigurationToConfigurationPlanner::Result` and `Planner::Result`? Maybe it'd be easier if we can always use `Planner::Result` so we don't have to change our result type if we decide we want to use a different planner? What would happen for a meta-planner?

jslee02 replied:
They are identical. Note that `SnapConfigurationToConfigurationPlanner` doesn't actually typedefs `Result`, but it just `Planner::Result`. Using `[X]Planner::Result` would be a safe way because it will always use the right result type.

Planners could have different types of information so it would be more flexible to define different result type when necessary.

Maybe we need to introduce `Planner::createResult()` to be able to create the right result from a base planner pointer like we did for `Outcome`.

brianhou replied:
Sounds good. Let's merge this PR first and add that change later.

aditya-vk commented on include/aikido/planner/ConfigurationToConfigurations.hpp:
```diff
@@ -0,0 +1,62 @@
+#define AIKIDO_PLANNER_CONFIGURATIONTOCONFIGURATIONS_HPP_
+
+#include <unordered_set>
+#include "aikido/constraint/Testable.hpp"
```
Not: There needs to be an empty line between aikido header and standard headers.

brianhou replied:
Why? `clang-format` won't order them correctly if we add empty lines.

aditya-vk replied:
From a recent comment by @psigen , AIKIDO follows this convention:
https://google.github.io/styleguide/cppguide.html#Names_and_Order_of_Includes

aditya-vk commented on include/aikido/planner/ConfigurationToEndEffectorOffset.hpp:
```diff
@@ -0,0 +1,77 @@
+  /// \param[in] startState Start state.
+  /// \param[in] direction Unit vector that represents the direction of motion
+  /// [unit vector in the world frame]. The direction is allowed to be zero
+  /// vector.
```
Should this docstring mention what direction being zero-vector means?

brianhou replied:
Maybe let's just throw an error if the direction is zero since we shouldn't ever actually want to do that.

aditya-vk replied:
Yeah.
@jslee02 why is it allowed to be zero vector?

aditya-vk commented on include/aikido/planner/Planner.hpp:
```diff
@@ -0,0 +1,75 @@
+#define AIKIDO_PLANNER_PLANNER_HPP_
+
+#include <string>
+
```
Uh oh :P

aditya-vk commented on include/aikido/planner/PlanningResult.hpp:
```diff
@@ -6,9 +6,22 @@
+  /// Message
+  std::string mMessage;
+
+  // TODO: Remove
```
Requires some work?

jslee02 replied:
This class will be removed anyway. :)

brianhou commented on src/planner/ConfigurationToEndEffectorOffset.cpp:
```diff
@@ -19,7 +19,8 @@ ConfigurationToEndEffectorOffset::ConfigurationToEndEffectorOffset(
   , mDistance(signedDistance)
 {
-  // Do nothing
+  if (direction.squaredNorm() == 0)
```
Should this be `direction.isZero()` instead?

aditya-vk commented on include/aikido/planner/SnapConfigurationToConfigurationPlanner.hpp:
```diff
@@ -0,0 +1,65 @@
+  /// \param[out] result Information about success or failure.
+  /// \return Trajectory or \c nullptr if planning failed.
+  /// \throw If \c problem is not ConfigurationToConfiguration.
+  /// \throw If \c result is not ConfigurationToConfiguration::Result.
```
If this splitting of `\throw` is the convention, can you please follow this even in `include/aikido/planner/Problem.hpp`. If there _is_ no convention, let us be consistent within the PR :)

aditya-vk commented on src/planner/FirstSupportedMetaPlanner.cpp:
```diff
@@ -0,0 +1,22 @@
+      return planner->plan(problem, result);
+  }
+
+  // TODO: Set result "no supported planner".
```
(bringing to your addition)

jslee02 replied:
I believe we will go over all the planner implementation including meta planners after this PR is merged.

aditya-vk commented on src/planner/Problem.cpp:
```diff
@@ -0,0 +1,45 @@
+  {
+    mConstraint = std::make_shared<constraint::Satisfied>(
+        std::const_pointer_cast<statespace::StateSpace>(mStateSpace));
+    // TODO(JS): Remove this const-cast once we fix const-correctness of
```
Can you reference this in #369, I'll take it up soon? Thanks!

aditya-vk commented on src/planner/SnapConfigurationToConfigurationPlanner.cpp:
```diff
@@ -0,0 +1,76 @@
+    mInterpolator
+        = std::make_shared<statespace::GeodesicInterpolator>(mStateSpace);
+    // TODO(JS): Not sure if GeodesicInterpolator is a good choice as the
+    // default.
```
Can you make an issue stating why? It can get lost here and seems important if there's valid reasons. Thanks!

jslee02 replied:
I put this comment without looking at other interpolators. Now I think it's a good choice. Let me delete this comments.

brianhou commented:
Thanks for the changes! Merging this now 🎉
